Kernel/Arch/i686/Bootloader/stivale2.c

```diff
@@ -0,0 +1,302 @@
+struct Table pagesPageTable;
+
+// page table to hold bootloader structures
+__attribute__((section(".early_bss")))
+struct Table bootPageTable;
+
+// stack for C/C++ runtime
+__attribute__((section(".early_bss")))
+struct Table stackPageTable;
+
+// bootloader info pointer (for preserving when stack changes)
+__attribute__((section(".early_bss")))
+struct stivale2_struct *stivale2Info;
+
+// stack for stage1 C/C++ runtime
+__attribute__((section(".early_bss")))
+uint8_t stage1Stack[KERNEL_STACK_SZ];
+
+// stack for stage2 C/C++ runtime
+__attribute__((section(".bss")))
+uint8_t stage2Stack[KERNEL_STACK_SZ];
+
+// TODO: Remove & rename once boot.s is gone
+uint32_t stivale2_mmap_helper(struct stivale2_tag* tag);
+
+//-----------------------------------------------
+// Stage1 functions
+//-----------------------------------------------
+
+static void stage1MapBootloader(void);
+static void stage1MapHighMemory(void);
+static void stage1MapLowMemory(void);
+static void stage1Entry(struct stivale2_struct *info);
+
+/**
+ * @brief Scan the stivale2 tags to find the reclaimable bootloader
+ * memory map tag that starts at the stivale2 tag base address. Return
+ * the length found. We'll memory map the entire area.
+ *
+ */
+__attribute__((section(".early_text")))
+uint32_t stivale2_mmap_helper(struct stivale2_tag* tag)
+{
+    while (tag) {
+        switch (tag->identifier) {
+            case STIVALE2_STRUCT_TAG_MEMMAP_ID:
+            {
+                struct stivale2_struct_tag_memmap* memmap = (struct stivale2_struct_tag_memmap*)tag;
+                for (size_t i = 0; i < memmap->entries; i++) {
+                    switch (memmap->memmap[i].type)
+                    {
+                        case STIVALE2_MMAP_BOOTLOADER_RECLAIMABLE:
+                            if (((uint32_t)memmap->memmap[i].base) == (uint32_t)tag) {
+                                return (uint32_t)memmap->memmap[i].length;
+                            }
+                            break;
+
+                        default:
+                            break;
+                    }
+                }
+            }
+            default:
+                break;
+        }
+        tag = (struct stivale2_tag*)(uint32_t)tag->next;
+    }
+    // If we get here, there's a problem so we will panic and never return
+    EarlyPanic("Error: Cannot detect bootloader info length!");
+}
+
+/**
+ * @brief Finalize stage1 by loading the finalized page directory
+ * and enabling paging.
+ *
+ */
+__attribute__((section(".early_text")))
+static void stage1EnablePaging(void)
+{
+    setPageDirectory((uintptr_t)&pageDirectory);
+
+    struct CR0 cr0 = readCR0();
+    cr0.pagingEnable = 1;
+    writeCR0(cr0);
+}
+
+/**
+ * @brief Prepare a higher-half stack for kernel usage.
+ *
+ */
+__attribute__((section(".early_text")))
+static void stage1JumpToStage2(void)
+{
+    // zero the kernel BSS (higher half stack)
+    for (size_t i = 0; i < _BSS_SIZE; i++) {
```

Comment:
You need to take the address of these symbols, otherwise you're just trying to access the value at that address.